### DISCUSSION TITLE
ReserveCalifornia Suggestion

### DISCUSSION THREAD

**Discussion by steffanvigano** [juftin/camply#238] [Ideas]
Hi again, 
It seems like Reserve California sites aren't searchable by specific campsite ID (number), is that correct?  Couple questions (these also apply to the Yosemite Reservations):

* Would it be possible to parse the campground for list of campsites using the API?  

Sadly, the campsite ID is not part of the URL and looking at the page source, I don't seem to be able to find it.  EX:  New Brighton State Beach:  https://www.reservecalifornia.com/Web/#!park/685/598    Our favorite sites are 73 and 75.

Something like:
     camply campsites \
    --provider ReserveCalifornia \
    --campground 598 \
    _--list_campsites_

* If not, would it be possible to at least filter down to the results to Premium Sites only?   

Not sure how they do it, by seems like WanderlingLabs has figured it out for this location. 

Thanks again!
-S

**Comment by juftin** [comment]
Looking at this closer, right now ReserveCalifornia doesn't support searching by campsite ID because I haven't found a way do perform a reverse lookup on campground ID by campsite ID. There are underlying IDs though, for example Campsite #47 @ that campground actually has ID #43440. 

```console
❯ camply campsites --debug \
    --campground 598 \
    --start-date 2023-07-01 \
    --end-date 2023-07-31 \
    --notifications silent

[2023-04-05 15:41:39] CAMPLY   camply, the campsite finder ⛺️
[2023-04-05 15:41:39] INFO     Using Camply Provider: "ReserveCalifornia"
[2023-04-05 15:41:39] DEBUG    Setting up camply debugging
[2023-04-05 15:41:39] DEBUG    Camply Version: 0.16.0
[2023-04-05 15:41:39] DEBUG    Python Version: 3.10.9
[2023-04-05 15:41:39] DEBUG    Platform: darwin
[2023-04-05 15:41:39] INFO     30 booking nights selected for search, ranging from 2023-07-01 to 2023-07-30
[2023-04-05 15:41:39] INFO     Searching for campsites every 10 minutes.
[2023-04-05 15:41:39] INFO     Notifications active via: <SilentNotifications>
[2023-04-05 15:41:39] INFO     Only <SilentNotifications> enabled. I hope you're watching these logs.
[2023-04-05 15:41:39] INFO     Searching across 1 campgrounds
[2023-04-05 15:41:39] INFO         ⛰  New Brighton SB (#685) - 🏕  Campground Northern End (sites 44-111) (#598)
[2023-04-05 15:41:39] INFO     Searching Campground Northern End (sites 44-111), New Brighton SB (598) for availability: July, 2023
[2023-04-05 15:41:39] DEBUG    Starting new HTTPS connection (1): calirdr.usedirect.com:443
[2023-04-05 15:41:39] DEBUG    https://calirdr.usedirect.com:443 "POST /rdr/rdr/search/grid HTTP/1.1" 200 None
[2023-04-05 15:41:39] INFO             ⛺️      1 total sites found in month of July
[2023-04-05 15:41:40] INFO     ⛺️ ⛺️ ⛺️ ⛺️ 1 Reservable Campsites Matching Search Preferences
[2023-04-05 15:41:40] INFO     📅 Mon, July 03 🏕  1 sites
[2023-04-05 15:41:40] INFO             ⛰️  New Brighton SB  🏕  Campground Northern End (sites 44-111): ⛺ 1 sites
[2023-04-05 15:41:40] INFO                     🔗 https://www.reservecalifornia.com/Web/Default.aspx#!park/685/598 (1 night)
[2023-04-05 15:41:40] INFO     1 New Campsites Found.
[2023-04-05 15:41:40] DEBUG    SilentNotification:
                                       • 2023-07-03 - 2023-07-04
                                       • Hook Up Camping
                                       • Hook Up (E/W) Campsite #47
                                       • New Brighton SB
                                       • Campground Northern End (sites 44-111)
                                       • https://www.reservecalifornia.com/Web/Default.aspx#!park/685/598
[2023-04-05 15:41:40] DEBUG    Campsite Info: {'availability_status': 'Available',
                                'booking_date': datetime.datetime(2023, 7, 3, 0, 0),
                                'booking_end_date': datetime.datetime(2023, 7, 4, 0, 0),
                                'booking_nights': 1,
                                'booking_url': 'https://www.reservecalifornia.com/Web/Default.aspx#!park/685/598',
                                'campsite_attributes': None,
                                'campsite_id': 43440,
                                'campsite_loop_name': None,
                                'campsite_occupancy': (0, 1),
                                'campsite_site_name': 'Hook Up (E/W) Campsite #47',
                                'campsite_type': 'Hook Up Camping',
                                'campsite_use_type': 'Hook Up (E/W)',
                                'facility_id': 598,
                                'facility_name': 'Campground Northern End (sites 44-111)',
                                'permitted_equipment': None,
                                'recreation_area': 'New Brighton SB',
                                'recreation_area_id': 685}
[2023-04-05 15:41:40] CAMPLY   Exiting camply 👋
```

**Comment by steffanvigano** [comment]
Thanks.  Now that I know about the debug flag, I can also see the campsite ID's.  Seems like they are in somewhat numerical order based on what I can see for availability for the whole summer, but do not increment in order even though there are campsites with those numbers:

    'campsite_id': 43450,, 'campsite_site_name': 'Campsite #57',   
    'campsite_id': 43453,, 'campsite_site_name': 'Campsite #61',  (seems like this should be 43454)
    'campsite_id': 43457,, 'campsite_site_name': 'Campsite #72',  (again this doesn't add up.. 61 to 72 should be more than 4)

So unfortunately it's not a slam dunk to extrapolate campsite_id's.

**Reply by juftin** [reply]
There are endpoints like https://calirdr.usedirect.com/rdr/rdr/fd/places that provide information about ReserveCalifornia's campgrounds and recreation areas. 

There is likely an endpoint to give us information about campsites  within campgrounds - this is what will be needed in order to build out the functionality to filter by campground IDs. 

Once that's found this will be quite straightforward to build out. Unfortunately, ReserveCalifornia blocks browser developer tools so it's extremely difficult to figure out what API endpoints the website is talking to.

**Reply by mcg88** [reply]
URL: https://calirdr.usedirect.com/RDR/rdr/search/grid
POST with the following body
```
{"IsADA":false,"MinVehicleLength":0,"UnitCategoryId":1,"StartDate":"2023-05-09","WebOnly":true,"UnitTypesGroupIds":[],"SleepingUnitId":0,"EndDate":"2023-05-16","UnitSort":"orderby","InSeasonOnly":true,"FacilityId":"597","RestrictADA":false}
```
Gets you a json file with all the campsites. You can get around teh dev tools restriction using this: https://stackoverflow.com/questions/49795510/disable-function-call-to-debugger-in-chrome-developer-console

**Reply by juftin** [reply]
Oh nice, thanks @mcg88. I'm already using [RDR/rdr/search/grid](https://calirdr.usedirect.com/RDR/rdr/search/grid) and parsing its entire response - I didn't even think of using it for building out a mapping of campsites. 

This is doable and decently straightforward, but it will require providing both the campground ID and the campsite ID as a reverse lookup using only campsite ID isn't yet possible.

**Comment by juftin** [comment]
Check out the latest list-campsites feature here: https://juftin.com/camply/command_line_usage/#list-campsites 🎉

**Comment by steffanvigano** [comment]
Much appreciated.  Works like a charm.

FYI that the command line switches of “campground” and “campsite” both require that you add an ‘S’ within the YAML config file.  So “campgrounds” and “campsites” vs singular in the command line.  That tripped me up a bit.  Perhaps that’s by design?
-S
